Paper Planes Mk2

Mercenaryblue opened the pull request:
Re-uploaded from a fresh repo branch

:cl:
rscadd: Adds a basic paper plane to the game. Fold paper with alt-clicking.
rscadd: Paper planes are now fully compatible with every stamps on station.
/:cl:
**Now its own object, with its own file**
Added sprites and overlay.
Added basic paper plane code.
**Crafted from paper with alt-click**
Can unfold back to paper with alt-click
**Compatible with stamps!**
**Doesn't fuck with cult talismans**
**Info transfers between states.**
May proc random eye damage by chance.
Fixed throw code.
Fixed burning code. 
Now prioritize talisman code when alt-click, no longer overriding.

Thanks to PKPenguin321 for general help and Eye Proc.
Thanks to KazeEspada for help with the new internal paper version
Thanks to JJRCop for the info transfer code
Thanks to Coderbus for general help and advice.

Mercenaryblue committed:
Fresh rep, fresh branch, no conflicts.

Re-upload code and sprites.

Chocobro commented:
:+1: also I don't know if we still need to wait 24h on account of the last pr being real for 5 days

Mercenaryblue commented:
Considering the amounts of mistakes made on #16970 , I'm fine with 24h to check everything out.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+	layer = 3
+	pressure_resistance = 0
+	burn_state = FLAMMABLE
+	burntime = 5
```
Throwforce, Gender, W_class, layer, pressure Resistance, and burn vars are all unnessary. They are all handled by the parents.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+
+/obj/item/weapon/paper/paperplane/extinguish()
+	..()
+	update_icon()
```
Isn't this the same as papers?

Mercenaryblue replied:
Codewise, yes. But it was needed so the plane didn't revert back to burning paper.

Mercenaryblue committed:
Remove un-needed variables

Handled by parent tree

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+		H.adjust_blurriness(6)
+		H.adjust_eye_damage(rand(6,8))
+		H.Weaken(2)
+		H.emote("scream")
```
Pretty sure this should be in throw_impact()?

Mercenaryblue replied:
oh f*cking hell this is the old throw code one moment

PKPenguin321 replied:
Remember to switch !..() to ..() when you switch the proc

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+	..()
+	if(!in_range(src, user))
+		return
+	if(!istype(src, /obj/item/weapon/paper/paperplane) && !istype(src, /obj/item/weapon/paper/talisman)) //don't fuck with cult
```
Override AltClick() on the children, don't do snowflake checks.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+/obj/item/weapon/paper/paperplane
+	name = "\improper paper plane"
+	desc = "paper folded in the shape of a plane"
+	icon = 'icons/obj/bureaucracy.dmi'
```
This was set by the parent.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+		return
+
+/obj/item/weapon/paper/paperplane
+	name = "\improper paper plane"
```
\improper is only needed for things that are capitalized but not proper nouns.

Mercenaryblue replied:
Fixed

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+			I = new /obj/item/weapon/paper/paperplane(src.loc)
+			user.put_in_hands(I)
+			src.loc = I
+			I.CheckParts()
```
This whole section is garbage.

Iamgoofball replied:
This is how it should be done.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+			src.stamped = P.stamped.Copy()
+		src.rigged = P.rigged
+		qdel(P)
+		updateinfolinks()
```
This code is pretty bad too.

Iamgoofball replied:
Nah, it's needed.

JJRcop replied:
so what's wrong with it?

KazeEspada replied:
Mostly the locate() in src. Like why? Instead of just holding a reference to the paper?

JJRcop replied:
tablecrafting

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+	if(!istype(src, /obj/item/weapon/paper/paperplane) && !istype(src, /obj/item/weapon/paper/talisman)) //don't fuck with cult
+		user << "<span class='notice'>You fold the paper in the shape of a plane!</span>"
+		if(do_after(user, 20, target = src))
+			user.unEquip(src)
```
This can return 0 if the paper is ABSTRACT or NODROP.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+		user << "<span class='notice'>You fold the paper in the shape of a plane!</span>"
+		if(do_after(user, 20, target = src))
+			user.unEquip(src)
+			get_turf(src)
```
get_turf is used to get a turf and store it somewhere? This effectively does nothing.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+			get_turf(src)
+			I = new /obj/item/weapon/paper/paperplane(src.loc)
+			user.put_in_hands(I)
+			src.loc = I
```
It should be src.forceMove(I).

Iamgoofball replied:
This is valid, forcemove is safer

JJRcop replied:
my fault

Mercenaryblue committed:
Push the Fixed Throw code.

Made by PKPenguin321

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,82 @@
+			I.CheckParts()
+		return
+
+/obj/item/weapon/paper/paperplane
```
TBH, I am not completely sure why this is a child of paper? I mean, you really can't read or write on a paper plane?

Mercenaryblue replied:
Sure you can. You need to unfold it IRL, but you can.

KazeEspada replied:
If you unfolded it... than it would be paper?

TheWiznard replied:
and then after you write on it you... fold it back into a plane??

TrustyGun commented:
fam you messed up your changelog

do it like this:

:cl:
rscadd: Memes
/:cl:

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,78 @@
+
+/obj/item/weapon/paper/paperplane/throw_at(atom/target, range, speed, mob/thrower, spin=0)
+	if(!..())
+		return
```
Is this still needed?

phil235 replied:
this doesn't do anything, you don't need it.

Mercenaryblue replied:
Roger. Next revision.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,78 @@
+	..()
+	if(!in_range(src, user))
+		return
+	if(!istype(src, /obj/item/weapon/paper/paperplane) && !istype(src, /obj/item/weapon/paper/talisman)) //doesn't fuck with cult
```
Override the altclick() on the children, don't do a snowflake check.

KazeEspada commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,78 @@
+		user << "<span class='notice'>You fold the paper in the shape of a plane!</span>"
+		if(do_after(user, 20, target = src))
+			user.unEquip(src)
+			get_turf(src)
```
get_turf is doing nothing here.

Mercenaryblue replied:
Will delete.

Mercenaryblue committed:
Fixes

Removed \improper
Removed get_turf(src)
Used drop_item
Fixed stacking sprites while on fire.

Mercenaryblue committed:
Force Move (I)

As suggested

Mercenaryblue commented:
Only need to fix the snowflake check

KazeEspada commented:
TBH, I would make the paper airplane its own object, and store the paper inside so that people have to unfold it to read and write it. Also, you could fold talismans with no problem so that you could send them to nearby cultists.

Mercenaryblue commented:
I was thinking of adding cultist paper planes functions later, yeah.
I just wanted to start with the basic item.

KazeEspada commented:
Ranged talismens?

Mercenaryblue commented:
That, or just easy passing

KazeEspada commented:
Actually upon further thought, I firmly believe that the planes should not be paper themselves and just store a reference to their source paper inside themselves.

Mercenaryblue commented:
Yikes. Okay. Let me close this while I prepare the independent paperplanes.

Mercenaryblue committed:
Moved /paperplane/, no longer paper child

It has its own file now.

Mercenaryblue committed:
Revert pen.dm

accidently killed it. my bad.

Mercenaryblue committed:
Info Transfer Fix

Overlays, Text, and Icon now transfer between states.

Mercenaryblue committed:
Dear @PKPenguin321

// Dear PKPenguin321
// We need to make sure this AltClick doesn't override the child objects
(aka talismans)
We need to redefine the AltClick on Child, Force Do Not Load, then
return.

This way alt-click craft will be unique to papers.

Your advice is appreciated.  Once this hurdle is passed we can
streamline the code.

Mercenaryblue commented:
@PKPenguin321 - Check out the branch. May need your advice there.

Mercenaryblue committed:
Checks dead/restrained status

Prevent folding paper when cuffed or dead.

Mercenaryblue commented:
Changes:
"weapon/paperplane/" is now its own object, with its own file.
Unfolding mechanism added. The paper retains the info it had before it was folded.
Fixed burning/extinguishing bug

Mercenaryblue committed:
Streamline code

Remove unused code

Mercenaryblue committed:
Revert "Streamline code"

This reverts commit 7900330681e72650444aa712e2e89c62b5ceeaec.

Mercenaryblue committed:
Revert "Revert "Streamline code""

This reverts commit 9b31867fdb1d249e498ceb80f3c5f8b1f3467c0b.

Mercenaryblue committed:
Prevent folding of unusual papers

Plus lays the groundwork toward talisman proc?

Mercenaryblue committed:
Talisman Alt-Click

Empty placeholder code.
Now the game will prioritize talisman code.

Mercenaryblue committed:
Talisman Placeholder Code

Until we get around to make cult paperplanes

Mercenaryblue committed:
Cleaning extra space

Mercenaryblue committed:
Head_Slot removed

can't believe I missed this.

KazeEspada committed:
Fixes Airplanes for the last time.

KazeEspada committed:
Merge branch 'paperplane' of https://github.com/Mercenaryblue/-tg-station into airplanes

Mercenaryblue committed:
Merge pull request #1 from KazeEspada/airplanes

Fixes Airplanes for the last time.

KazeEspada committed:
Update talisman.dm

Mercenaryblue commented:
Looks like we'll need new overlay sprites.
I'll do them tomorrow.

JJRcop commented:
This code looks really good now, you seem to have learned a lot..
..actually as I say that I'm looking at @KazeEspada's commit and it looks like he did most of it, uh.. Well I hope you look at e882b208f403566723ee07f1b3d5279fda87fab8 for a bit, and maybe you'll learn stuff!

Mercenaryblue commented:
I've got like one week of experience, tops. I'm amazed my stuff even works.

JJRcop commented:
Yeah, that is absolutely something to be happy about. I definitely recommend diving into the code of some features of SS13 you like and are intrigued with, then that will inevitably lead you somewhere else, then somewhere else. I think understanding the flow of the click code is pretty interesting, for example. 

And keep making stuff, don't stop!

KazeEspada commented:
His framework was solid, there was just a lot of code bloat and weird ways of doing things.

Also, I removed the do_after(), I allowed talismans to become airplanes, although airplanes cant be read without unfolding. I moved unfolding to just clicking it in hand without any modifiers. Removed the weirdness on talismans since none of the talismans use alt-click.

Mercenaryblue commented:
Once I upload them, the stamps will look like this on the paper planes.
![stamps](https://cloud.githubusercontent.com/assets/6573763/14769311/540103be-0a24-11e6-9b9e-0aa66de8c76a.png)

phil235 commented:
Squash your commits please.

Iamgoofball commented:
@phil235 Not needed anymore, you can squash & merge

Mercenaryblue committed:
Paperplane stamps

* First batch
Captain, HoS, HoP, CE, RD, CMO, Clown, QM, Cent Stamps included.

* Second batch
Law, Okay, and Deny stamps included.

Stamping Code 1.1.0
Paper.dm Line 305 removed at the request of coderbus.
Final stamping code. No more bugs.

Mercenaryblue commented:
For future reference, how do I squash?

PKPenguin321 commented:
you really need to not close and reopen this so much

Mercenaryblue committed:
Squash paperplane

Mercenaryblue commented:
What I googled told me not to squash while open, so.

I'll keep it in mind next PR

PKPenguin321 commented:
yeah it's cool it's just that every time you close it and open it it sends a message to the server and the IRC, it can get spammy

Mercenaryblue commented:
I'm not sure the squash worked. Ah well. (Maybe I needed to keep the first one at pick)

Mercenaryblue committed:
Delete heiz4blj.r0j.txt

KazeEspada commented:
Just have the maintainer squash during merge.

Mercenaryblue commented:
Yeah. Having an option to easily merge your own commit would be nice too.

Mercenaryblue commented on code/modules/paperwork/paper.dm:
```diff
@@ -302,11 +304,11 @@
 			return
 
 	else if(istype(P, /obj/item/weapon/stamp))
-		if(!in_range(src, usr) && loc != user && !istype(loc, /obj/item/weapon/clipboard) && loc.loc != user && user.get_active_hand() != P)
```
Anyway. I was told by coderbus to remove this line. They said it'll never trigger. 
I have doubts about removing code I didn't write myself, so I will gladly re-instate it if asked.

Supermichael777 commented:
does the eye damage respect eyepoke protection?

PKPenguin321 commented:
yes

phil235 commented on code/modules/paperwork/paper.dm:
```diff
@@ -28,6 +28,7 @@
 	var/list/stamped
 	var/rigged = 0
 	var/spam_flag = 0
+	//var/stamptype = "paper" //define what category of stamps we're using
```
don't leave commented code.

phil235 commented on code/modules/paperwork/paper.dm:
```diff
@@ -302,11 +304,11 @@
 	else if(istype(P, /obj/item/weapon/stamp))
-		if(!in_range(src, usr) && loc != user && !istype(loc, /obj/item/weapon/clipboard) && loc.loc != user && user.get_active_hand() != P)
+
+		if(!in_range(src, usr))
```
"user", not "usr"

phil235 commented:
Apart from the two comments I made above, the PR is good to go. :+1:

Mercenaryblue committed:
Fix User

Also remove commented code

PKPenguin321 commented:
:hooray:
